specs/deneb/fork-choice.md

```diff
@@ -23,54 +22,42 @@ This is the modification of the fork choice accompanying the Deneb upgrade.
 
 ## Containers
 
-### `BlobsSidecar`
-
-```python
-class BlobsSidecar(Container):
-    beacon_block_root: Root
-    beacon_block_slot: Slot
-    blobs: List[Blob, MAX_BLOBS_PER_BLOCK]
-    kzg_aggregated_proof: KZGProof
-```
-
 ## Helpers
 
-#### `validate_blobs_sidecar`
+#### `validate_blob_sidecars`
 
 ```python
-def validate_blobs_sidecar(slot: Slot,
+def validate_blob_sidecars(slot: Slot,
                            beacon_block_root: Root,
                            expected_kzg_commitments: Sequence[KZGCommitment],
-                           blobs_sidecar: BlobsSidecar) -> None:
+                           blob_sidecars: Sequence[BlobSidecar]) -> None:
     assert slot == blobs_sidecar.beacon_block_slot
     assert beacon_block_root == blobs_sidecar.beacon_block_root
-    blobs = blobs_sidecar.blobs
-    kzg_aggregated_proof = blobs_sidecar.kzg_aggregated_proof
-    assert len(expected_kzg_commitments) == len(blobs)
-
-    assert verify_aggregate_kzg_proof(blobs, expected_kzg_commitments, kzg_aggregated_proof)
+    assert len(expected_kzg_commitments) == len(blob_sidecars)
+    # TODO validate commitments individually or aggregate first?
+    # assert verify_aggregate_kzg_proof(blobs, expected_kzg_commitments, kzg_aggregated_proof)
```

Comment:
Okay so we're going to use `verify_blob_kzg_multi()` (see [here](https://github.com/ethereum/consensus-specs/pull/3236/files#diff-44e74c6479198ca2d57ab8c0895785dc5ab016145e9472de9bdcd94ab105265bR509))

This needs -- `(blobs, commitments, proofs)` so
I suggest that the fork-choice document does not bring *any* of the networking data--structures -- `BlobSidecar` or `AggregateBlobSidecar` and instead just assumes `retrieve_blobs` returns a tuple of `(blobs, proofs)`. This will nicely prevent forkchoice from having to care about these networking structures that are different depending on retrieval message (gossip vs byroot vs byrange)

So concretely:
* rename `validate_blob_sidecars` to `validate_blobs` or `validate_blob_commitments`
* rename `retrieve_blob_sidecars` to `retrieve_blobs_and_proofs` which returns a tuple of `(blobs, proofs)`
* In `is_data_available`, set `blobs, proofs = retrieve_blobs_and_proofs(slot, beacon_block_root)`
* `validate_blobs` function signature is `(slot, beacon_block_root, expected_kzg_commitments, blobs, proofs)` instead of  using `blob_sidecars`
* `validate_blobs` calls `verify_blob_kzg_proof_multi(blobs, expected_kzg_commitments, proofs)`

Reply:
A client that wants to iteratively verify commitments can do so with `_multi()` or with the a specific single commitment verification function. 

Nonetheless, I see that as an implementation detail. The *fork-choice* spec says "for the entire block, did we get the blobs, the correct number of commitments, and are they valid". This might happen iteratively with gossip, or in aggregate with some historic sync. But fc spec doesn't need to care about the different network delivery options

Reply:
The batch-verification function is [verify_blob_kzg_proof_multi()](https://github.com/ethereum/consensus-specs/pull/3236/files#diff-44e74c6479198ca2d57ab8c0895785dc5ab016145e9472de9bdcd94ab105265bR509) and the single blob verification function is [verify_blob_kzg_proof()](https://github.com/ethereum/consensus-specs/pull/3236/files#diff-44e74c6479198ca2d57ab8c0895785dc5ab016145e9472de9bdcd94ab105265bR509).

Using the former will give a reasonable speed boost especially as the number of blobs grows.

If you want a single generic function that invokes the multi function or the single function depending on the number of blobs this might be a good time to introduce it on the crypto layer (or it can also be done on the layer above).

Reply:
sgtm, specially the part about dropping networking types, though:

> validate_blobs calls verify_blob_kzg_proof_multi(blobs, expected_kzg_commitments, proofs)

While it's excellent that we have access to an efficient function for multiple blobs, I think we might be overconstraining in calling it here compared to other portions of the spec. For example, clients batch-verify signatures of all attestations in a block, but the spec still treats them independently for purpose of validation and validates one-by-one in the spec text, because this is the most simple way to express the spec requirement.

The above is interesting for two reasons: when importing blocks and blobs from sync, it's likely we'll want to batch-validate blobs from several blocks in one call, while at the same time, we want to sometimes validate a single blob when the client is not doing anything else - this freedom is somewhat .. muddled when introducing the multi call here.

Semantically, there's not much difference really, except that "batching" is slightly more complex to reason about (you don't find out which blob is bad, so you need to verify them individually then potentially try to recover etc).

Another way to put it: the minimal requirement from a spec perspective is that we have to check each blob, so lets maybe write that in the spec with a loop of single-proof-verifications and leave the multi-optimization implementation-specific.

Reply:
> If you want a single generic function that invokes the multi function or the single function depending on the number of blobs this might be a good time to introduce it on the crypto layer (or it can also be done on the layer above).

I don't think this belongs in the spec, really - it's good to have both the most efficient "single-verification" function and a "multi-verification" function, but choosing between them can remain an implementation matter.

See comment above however: in the spec text in particular, there might be reason to use the single-verification function throughout - compound operations like multi-verification are typically harder to reason about.

Another point to consider: it might be that blocks have only one blob - I'm assuming calling the multi function is slower for a single blob verification? What about 2 blobs? what's the performance profile? since these fine points may change over time, expressing the requirement in the simplest operations possible is imo the way to go.

Reply:
Agree, introducing `verify_blob_kzg_proof_multi()` looks to me like pulling implementation specifics into spec

Reply:
> Another point to consider: it might be that blocks have only one blob - I'm assuming calling the multi function is slower for a single blob verification? What about 2 blobs? what's the performance profile? since these fine points may change over time, expressing the requirement in the simplest operations possible is imo the way to go.

You can see some benchmarks in https://github.com/ethereum/c-kzg-4844/pull/123 . In general, seems like calling the multi function for a single blog (`VerifyBlobKZGProofMulti(blobs=1)`) is about 5% slower than verifying it with the single function (`VerifyBlobKZGProof`). Calling the multi function for two blobs is  already faster than calling the single function twice.

Reply:
Yeah, I'm totally fine with just iterating over `verify_blob_kzg_proof()` here and leaving the actual call depending on context to the implementer

Reply:
this came out more simple than expected: da34af97d4b6575131e67c2bf22acb6de4f0951d

In particular, since we're no longer verifying sidecars, we don't need to check slots etc.